### PR TITLE
First implementation of real-time slack computing

### PR DESCRIPTION
**Pull request by Peter-JanGootzen**
https://github.com/Peter-JanGootzen/vu-ds-energy-aware-scheduling/issues/4

Needs to be moved to JobInstance and TaskInstance, as they have the actual arrival time of the Job and Task.

### CODE REVIEW

File: opendc-workflow/opendc-workflow-service/src/main/kotlin/org/opendc/workflow/service/internal/StateHelpers.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import kotlin.collections.HashSet
+import kotlin.math.max
+
+public fun JobState.computeSlack() {
+    this.computeMinimalStartTimes()
+    for (t in this.tasks) {
+        if (t.dependencies.isEmpty())  {
+            t.task.metadata[WORKFLOW_TASK_SLACK] = 0L
+        } else {
+            val cmin = t.dependencies.minWith(Comparator.comparingLong {it.task.metadata[WORKFLOW_TASK_SLACK] as Long})
```
Comment (bwesterveld) [inline opendc-workflow/opendc-workflow-service/src/main/kotlin/org/opendc/workflow/service/internal/StateHelpers.kt]: The IDE doesn't give an error, but it.task.metadata does not exist and will result in a NULL pointer exception at runtime.

Comment (bwesterveld) [inline opendc-workflow/opendc-workflow-service/src/main/kotlin/org/opendc/workflow/service/internal/StateHelpers.kt]: After splitting it out into the following lines, I still get a runtime exception from casting a null value to long.

```  for (t in this.tasks) {
        if (t.dependencies.isEmpty())  {
            t.task.metadata[WORKFLOW_TASK_SLACK] = 0L
        } else {
            val cmin = t.dependencies.minWith(Comparator.comparingLong {t.task.metadata[WORKFLOW_TASK_SLACK]?: 0L as Long})
            val taskslack: Long = cmin.task.metadata[WORKFLOW_TASK_SLACK]?: 0L
            val minstarttime: Long = t.task.metadata[WORKFLOW_TASK_MINIMAL_START_TIME]?: 0L // exception here
            val taskdeadline: Long = t.task.metadata[WORKFLOW_TASK_DEADLINE]?: 0L
            t.task.metadata[WORKFLOW_TASK_SLACK] = taskslack - minstarttime + taskdeadline
        }
    
```

**Comment by Peter-JanGootzen** [comment]
My brain assumed that the Task dependencies would also be in topological order. They are of course not. So now I default the slack to 0, which is what it should be from the algorithm.

**Comment by Peter-JanGootzen** [comment]
This PR has been merged into the branch TaskFlow

**Pull request closed without merging**